### ISSUE TITLE
Got "UnicodeEncodeError: 'ascii' codec can't encode characters ..." using MailSender

### ISSUE THREAD

**Issue by runbing** [scrapy/scrapy#3710]
I want to send a message with MailSender when the spider closed. Here is the code snip:

```
# -*- coding:utf-8 -*-

import scrapy
from scrapy.mail import MailSender

class UpdateSpider(scrapy.Spider):
    name = 'update'

    # ...

    def close(self):
        mailer = MailSender.from_settings(self.settings)
        body = u'测试内容。'
        return mailer.send(to=["runbing@example.com"], subject=u"测试标题",
               body=body, mimetype="text/html")
```

but if the parameter **body**  is non-ascii string, I will get a error like blow.

```
2019-03-31 22:03:59 [scrapy.utils.signal] ERROR: Error caught on signal handler: <bound method ?.close of <UpdateSpider 'update' at 0x7f4932496650>>
Traceback (most recent call last):
  File "/home/runbing/Code/Scrapy/venv/lib/python2.7/site-packages/twisted/internet/defer.py", line 151, in maybeDeferred
    result = f(*args, **kw)
  File "/home/runbing/Code/Scrapy/venv/lib/python2.7/site-packages/pydispatch/robustapply.py", line 55, in robustApply
    return receiver(*arguments, **named)
  File "/home/runbing/Code/Scrapy/amazon/amazon/spiders/update.py", line 81, in close
    mimetype="text/html")
  File "/home/runbing/Code/Scrapy/venv/lib/python2.7/site-packages/scrapy/mail.py", line 101, in send
    dfd = self._sendmail(rcpts, msg.as_string().encode(charset or 'utf-8'))
  File "/usr/lib64/python2.7/email/message.py", line 137, in as_string
    g.flatten(self, unixfrom=unixfrom)
  File "/usr/lib64/python2.7/email/generator.py", line 83, in flatten
    self._write(msg)
  File "/usr/lib64/python2.7/email/generator.py", line 108, in _write
    self._dispatch(msg)
  File "/usr/lib64/python2.7/email/generator.py", line 134, in _dispatch
    meth(msg)
  File "/usr/lib64/python2.7/email/generator.py", line 180, in _handle_text
    self._fp.write(payload)
UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-1: ordinal not in range(128)
```

I spend all my day to solve this problem, but get no luck, I just find a way like below to hack the problem, but I know this is not recommended. 

```
import sys
reload(sys)
sys.setdefaultencoding('utf-8')
```
Is there any help for this problem?

**Comment by Gallaecio** [comment]
Have you tried using `body.encode('utf8')` in your `mailer.send` call?

**Comment by runbing** [comment]
> Have you tried using `body.encode('utf8')` in your `mailer.send` call?

Yes, I have tried it. But I got same error with changed error message when I use ```body.encode('utf8')``` in ```mailer.send```.

```
2019-04-01 10:13:44 [scrapy.utils.signal] ERROR: Error caught on signal handler: <bound method ?.close of <UpdateSpider 'update' at 0x7f84bb586650>>
Traceback (most recent call last):
  File "/home/runbing/Code/Scrapy/venv/lib/python2.7/site-packages/twisted/internet/defer.py", line 151, in maybeDeferred
    result = f(*args, **kw)
  File "/home/runbing/Code/Scrapy/venv/lib/python2.7/site-packages/pydispatch/robustapply.py", line 55, in robustApply
    return receiver(*arguments, **named)
  File "/home/runbing/Code/Scrapy/amazon/amazon/spiders/update.py", line 80, in close
    body=body, mimetype="text/html")
  File "/home/runbing/Code/Scrapy/venv/lib/python2.7/site-packages/scrapy/mail.py", line 101, in send
    dfd = self._sendmail(rcpts, msg.as_string().encode(charset or 'utf-8'))
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe4 in position 167: ordinal not in range(128)
```

**Comment by Gallaecio** [comment]
That traceback says `body=body, mimetype="text/html")`.

I was expecting to see `body=body.encode('utf8'), mimetype="text/html")` instead. Are you sure that traceback includes the encoding of the body?

**Comment by Gravellent** [comment]
I tested the code myself using gmail client and it works.

**Comment by runbing** [comment]
> That traceback says `body=body, mimetype="text/html")`.
> 
> I was expecting to see `body=body.encode('utf8'), mimetype="text/html")` instead. Are you sure that traceback includes the encoding of the body?

Sorry I didn't make it clear. I did add `encode('utf-8')` to body parameter but `mailer.send` call. If I add `encode('utf-8')` in `mailer.send` call, I got the same error.

```
2019-04-02 12:30:14 [scrapy.utils.signal] ERROR: Error caught on signal handler: <bound method ?.close of <UpdateSpider 'update' at 0x7fc1ee0ab610>>
Traceback (most recent call last):
  File "/home/runbing/Code/Scrapy/venv/lib/python2.7/site-packages/twisted/internet/defer.py", line 151, in maybeDeferred
    result = f(*args, **kw)
  File "/home/runbing/Code/Scrapy/venv/lib/python2.7/site-packages/pydispatch/robustapply.py", line 55, in robustApply
    return receiver(*arguments, **named)
  File "/home/runbing/Code/Scrapy/amazon/amazon/spiders/update.py", line 81, in close
    body=body.encode('utf-8'), mimetype="text/html", charset="utf-8")
  File "/home/runbing/Code/Scrapy/venv/lib/python2.7/site-packages/scrapy/mail.py", line 101, in send
    dfd = self._sendmail(rcpts, msg.as_string().encode(charset or 'utf-8'))
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe4 in position 218: ordinal not in range(128)
```

---

I spent a day try to figure out this problem, And I found this problem related to the python version.

If I use python 3.6.6, just add `encode('utf-8')` to body parameter, everything is good. But If I  do same thing on python 2.7.5, I got the UnicodeEncodeError.

I try to modify the related Scrapy source code in `scrapy/mail.py` to this:

```
# ...

class MailSender(object):

    # ...

    def send(self, to, subject, body, cc=None, attachs=(), mimetype='text/plain', charset=None, _callback=None):

       # ...

        # if charset:
        #     msg.set_charset(charset)

        if attachs:
            msg.attach(MIMEText(body, 'plain', charset or 'us-ascii'))
            for attach_name, mimetype, f in attachs:
                part = MIMEBase(*mimetype.split('/'))
                part.set_payload(f.read(), charset)
                Encoders.encode_base64(part)
                part.add_header('Content-Disposition', 'attachment', filename=attach_name)
                msg.attach(part)
        else:
            msg.set_payload(body, charset)

        # ...
```

The UnicodeEncodeError disappeared and the code works on both python 2.7.5 & 3.6.6. But can not pass unittest.

```
..F..F
======================================================================
FAIL: test_send_attach_utf8 (__main__.MailSenderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "../tests/test_mail.py", line 110, in test_send_attach_utf8
    self.assertEqual(msg.get_charset(), Charset('utf-8'))
AssertionError: None != utf-8

======================================================================
FAIL: test_send_utf8 (__main__.MailSenderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "../tests/test_mail.py", line 87, in test_send_utf8
    self.assertEqual(msg.get_payload(), body)
AssertionError: '5L2g5aW9\n' != u'\u4f60\u597d'

----------------------------------------------------------------------
Ran 6 tests in 0.001s

FAILED (failures=2)
```

**Comment by runbing** [comment]
> I tested the code myself using gmail client and it works.

Which python version are you using?

**Comment by meteFANS** [comment]
3.7-------- 原始邮件 --------主题：Re: [scrapy/scrapy] Got "UnicodeEncodeError: 'ascii' codec can't encode characters ..." using MailSender (#3710)发件人：Runbing 收件人：scrapy/scrapy 抄送：Subscribed 
I tested the code myself using gmail client and it works.

Which python version are you using?

**Comment by runbing** [comment]
These days I tried to fix the problem I encountered by modifying the source code, but the changed code behaved differently in different Python versions. I think the current code has compatibility issues, but I still not able to solve this problem. I am looking forward that someone can solve these problems:

1. Got UnicodeEncodeError when the parameter body is non-ascii string in python 2.7.X.
2. Displaying wrong attachment filename when the parameter filename is non-ascii string.
